### PR TITLE
FEAT: API to batch update discord nicknames

### LINKED ISSUE #1452: Update Discord nicknames with rds usernames
# Create API endpoint to update Discord usernames for backend users

## Description
Currently, there is a need to synchronize the usernames of rds users in our Discord server with the usernames they choose for the backend. To achieve this, we need to create a new API endpoint in our backend, which can only be accessed by super users. This API will run a script that utilizes an external API to update the Discord usernames.

## Acceptance Criteria
1. Create a new API endpoint in the backend that can be accessed only by super users.
2. The API should send a PATCH request to the following endpoint: `rdsServerlessBaseUrl/guild/member`.
3. The request headers should include an `Authorization` field with the value `Bearer <Auth Token>` for authentication.
4. The request body should include the following parameters:
   - `discordId`: The Discord ID of the user whose username needs to be updated.
   - `userName`: The RDS username of the user.
5. When the super user runs this script, the backend should check if the user document contains a `discordId`.
6. If the `discordId` is present, the backend should call the external API mentioned above, passing the `discordId` and the corresponding `userName` to update the user's nickname in the Discord server.
7. The API should handle the following responses from the external API:
   - If the response contains the string "🚫 Bad Request Signature," it should be treated as a missing signature/auth header error.
   - If the response contains the string "Oops! We have encountered an internal Server Error," it should be treated as an internal server error.
   - If the response contains the string "User nickname changed successfully," it should be treated as a successful update.
   - In case the API returns any other response, it should be handled appropriately as well.
8. Since there are no status codes provided by the external API, we will rely on the response strings mentioned above for error and success detection.
9. Proper error handling should be implemented for different scenarios, and appropriate messages should be logged.
10. Unit tests should be created to verify the functionality of the new API endpoint and the script.

Please ensure that the external API (`discord-slash-commands` repository) mentioned in the acceptance criteria is properly set up and functional before integrating it with the backend. Additionally, implement super user authorization for accessing the new API endpoint to restrict it to authorized users only.

### PR DESCRIPTION
**Pull request by kotesh-arya**
### Issue
Closes #1452 

### Description:
Currently, there is a need to synchronize the usernames of rds users in our Discord server with the usernames they choose for the backend. To achieve this, we need to create a new API endpoint in our backend, which can only be accessed by super users. This API will run a script that utilizes an external API to update the Discord usernames.

###  Changes made:
- Added a API route to update the Nicknames of users in RDS discord server which runs a script.
- The API sends a PATCH request to the following endpoint: `rdsServerlessBaseUrl/guild/member`.
- The request headers should include an `Authorization` field with the value `Bearer <Auth Token>` for authentication.

### Testing:
- The script and API endpoint have undergone comprehensive testing to ensure accurate identification of users with a linked Discord account and proper updating of their nickname with the username in RDS-backend. Testing includes various scenarios, such as users who are verified and unverified, to ensure the script behaves as expected.

### Is Development Tested?

- [x] Yes
- [ ] No

### Test coverage stats:
CI-test-stats-nickname-aync-api

#### Before:
nicknames-before-batch-update

#### After:
nickname-batch-update-later

### API working process:
[Screencast from 05-09-23 05:03:53 PM IST.webm](https://github.com/Real-Dev-Squad/website-backend/assets/69259490/417d2dbd-b74d-4d5b-8656-936d13539045)

### CODE REVIEW

**Review by iamitprakash — CHANGES_REQUESTED** [review]

**Review by iamitprakash — APPROVED** [review]

**Comment by iamitprakash** [comment]
@prakashchoudhary07  can you verify API name please

**Comment by prakashchoudhary07** [comment]
> @prakashchoudhary07 can you verify API name please

We should come up with something better

The current name `POST users/discord/nickname`, i think it indicates that this API is adding/creating a nickname for users in discord, or for nickname for discord and it doesn't exactly indicate that we are performing a batch request

**Comment by kotesh-arya** [comment]
> > @prakashchoudhary07 can you verify API name please
> 
> We should come up with something better, i think
> 
> The current name `POST users/discord/nickname`, indicates that this API is adding/creating a nickname for users in discord, or for nickname for discord and it doesn't exactly indicate that we are performing a batch request

Thanks for pointing out @prakashchoudhary07 
I have chosen an alternative for this as POST `users/discord-server/nicknames` 
please let me know if this still doesn't meet the acceptance criteria
thank you.

File: controllers/users.js [hunk 1]
```diff
@@ -24,6 +24,40 @@ const {
...
 } = require("../constants/users");
 
+/**
+ * Patch Update user nickname
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const updateDiscordNicknames = async (req, res) => {
```
Comment (iamitprakash) [inline controllers/users.js:34]: discordId and username properties are both declared as const, which means that their values cannot be changed. However, the setUserDiscordNickname function is trying to change the value of the discordId property.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: controllers/users.js (hunk 1 above)
Comment (kotesh-arya) [reply controllers/users.js:34]: In the **setUserDiscordNickname** function, I am passing the **discordId** through the request body
   `body: JSON.stringify({ userName, discordId })` but not changing it.

File: test/integration/users.test.js
Comment (vinit717) [inline test/integration/users.test.js:1]: Can you write a test to check whether nickname is update to username as you are only checking API response

Comment (kotesh-arya) [reply test/integration/users.test.js:1]: I have added an new `expect` statement in the test-case of discord-service to check the updating of nicknames 👍🏻

**Review by iamitprakash — APPROVED** [review]

**Review by yesyash — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**